File: iosApp/iosApp/Presentation/views/Settings/Components/RulesetGrid.swift
//
//  RulesetGrid.swift
//  iosApp
//
//  Created by Himanshu Sherkar on 22/05/26.
//

import SwiftUI
import SharedLogic

struct RulesetGrid: View {
    let selected: ChessRuleset?
    let onSelect: (ChessRuleset) -> Void
    
    private let items: [(ChessRuleset, String, String)] = {
        let factories: [() -> ChessRuleset] = [
            ChessRuleset.Blitz.init,
            ChessRuleset.Quick.init,
            ChessRuleset.Action.init
        ]
        
        return factories.map { make in
            let rs = make()
            return (rs, "\(rs.name) chess", "\(rs.minutes)m")
        }
    }()
    
    var body: some View {
        HStack(spacing: 8) {
            ForEach(items, id: \.0) { ruleset, label, time in
                PresetCard(
                    type: label,
                    time: time,
                    isSelected: selected == ruleset,
                    onSelect: { onSelect(ruleset) }
                )
            }
        }
    }
}

private struct PresetCard: View {
    let type: String
    let time: String
    let isSelected: Bool
    let onSelect: () -> Void
    
    var borderColor: Color {
        withAnimation {
            isSelected ? Color.accentGreen : Color.border
        }
    }
    
    var bgColor: Color {
        withAnimation {
            isSelected ? Color(hex: "#0D1F0D") : Color.surface
        }
    }
    
    var textColor: Color {
        withAnimation {
            isSelected ? Color.accentGreen : Color.textPrimary
        }
    }
    
    var body: some View {
        VStack(spacing: 4) {
            Text(type.uppercased())
                .font(.system(size: 9))
                .foregroundColor(Color.textSecondary)
                .tracking(1.0)
            
            Text(time)
                .font(.system(size: 22, weight: .light))
                .foregroundColor(textColor)
        }
        .frame(maxWidth: .infinity)
        .padding(10)
        .background(bgColor)
        .clipShape(RoundedRectangle(cornerRadius: 10))
        .overlay(
            RoundedRectangle(cornerRadius: 10)
                .stroke(borderColor, lineWidth: 1)
        )
        .onTapGesture { onSelect() }
    }
}

File: androidApp/src/main/kotlin/com/kingslayer06/clockapp/presentation/views/settings/components/RulesetGrid.kt
package com.kingslayer06.clockapp.presentation.views.settings.components

import androidx.compose.animation.animateColorAsState
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.kingslayer06.clockapp.core.ui.ColorAccentGreen
import com.kingslayer06.clockapp.core.ui.ColorBorder
import com.kingslayer06.clockapp.core.ui.ColorSurface
import com.kingslayer06.clockapp.core.ui.ColorTextPrimary
import com.kingslayer06.clockapp.core.ui.ColorTextSecondary
import com.kingslayer06.clockapp.domain.models.ChessRuleset

@Composable
fun RulesetGrid(
    selected: ChessRuleset?,
    onSelect: (ChessRuleset) -> Unit
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        listOf(
            Triple(ChessRuleset.Blitz, "${ChessRuleset.Blitz.name} chess", "${ChessRuleset.Blitz.minutes}m"),
            Triple(ChessRuleset.Quick, "${ChessRuleset.Quick.name} chess", "${ChessRuleset.Quick.minutes}m"),
            Triple(ChessRuleset.Action, "${ChessRuleset.Action.name} chess", "${ChessRuleset.Action.minutes}m"),
        ).forEach { (ruleset, label, time) ->
            PresetCard(
                modifier = Modifier.weight(1f),
                type = label,
                time = time,
                isSelected = selected == ruleset,
                onClick = { onSelect(ruleset) }
            )
        }
    }
}

@Composable
private fun PresetCard(
    modifier: Modifier,
    type: String,
    time: String,
    isSelected: Boolean,
    onClick: () -> Unit
) {
    val borderColor by animateColorAsState(
        targetValue = if (isSelected) ColorAccentGreen else ColorBorder,
        label = "presetBorder"
    )
    val bgColor by animateColorAsState(
        targetValue = if (isSelected) Color(0xFF0D1F0D) else ColorSurface,
        label = "presetBg"
    )
    val textColor by animateColorAsState(
        targetValue = if (isSelected) ColorAccentGreen else ColorTextPrimary,
        label = "presetTextColor"
    )

    Surface(
        modifier = modifier
            .clip(RoundedCornerShape(10.dp))
            .clickable(onClick = onClick),
        color = bgColor,
        shape = RoundedCornerShape(10.dp),
        border = BorderStroke(1.dp, borderColor)
    ) {
        Column(
            modifier = Modifier.padding(10.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            Text(
                text = type.uppercase(),
                color = ColorTextSecondary,
                fontSize = 9.sp,
                letterSpacing = 1.sp
            )
            Text(
                text = time,
                color = textColor,
                fontSize = 22.sp,
                fontWeight = FontWeight.Light,
                letterSpacing = (-0.5).sp
            )
        }
    }
}